Add OIDC Auth0 extended tutorial

sberyozkin opened the pull request:
This PR adds an extended OIDC Auth0 tutorial, originally it was meant to be a blog post but after discussing with @maxandersen we agreed to make it a tutorial.

This is not a typical tutorial because there is no quickstart, it is meant to show how both code flow and bearer Auth0 tokens can be managed, how Quarkus OIDC can deal with Auth0 specific constraints, how to adapt the code etc.

This tutorial is meant to be the first one in the series dedicated to specific providers, as opposed to simple quickstarts showing some OIDC features in isolation

sberyozkin committed:
Add OIDC Auth0 extended tutorial

sberyozkin commented:
Converting to draft until #36610 is merged - as this PR misses some integration testing instructions - we cover how we can use Dev UI to test but a bit more content is needed re the testing

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,587 @@
+
+== References
+* xref:security-overview.adoc[Quarkus Security overview]
+* xref:security-oidc-code-flow-authentication.adoc[OIDC code flow mechanism for protecting web applications]
```
Would it add value to link to the well-known OIDC providers topic, like so:

```suggestion
* xref:security-oidc-code-flow-authentication.adoc[OIDC code flow mechanism for protecting web applications]
* xref:security-openid-connect-providers.adoc[Configuring well-known OpenID Connect providers]
```

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,587 @@
+[id="security-oidc-auth0-tutorial"]
+= Protect Quarkus web application by using Auth0 OpenID Connect provider
```
Missing articles in title.
```suggestion
= Protect a Quarkus web application by using the Auth0 OpenID Connect provider
```

sberyozkin replied:
@michelle-purcell Hi Michelle, this suggestion won't work, this is actually referring to an OIDC provider named `Auth0`

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,587 @@
+:diataxis-type: tutorial
+:categories: security,web
+
+xref:security-architecture.adoc[Quarkus Security] provides comprehensive OpenId Connect (OIDC) and OAuth2 support with its `quarkus-oidc` extension, supporting both xref:security-oidc-code-flow-authentication.adoc[Authorization code flow] and xref:security-oidc-bearer-token-authentication.adoc[Bearer token] authentication mechanisms.
```
```suggestion
xref:security-architecture.adoc[Quarkus Security] provides comprehensive OpenID Connect (OIDC) and OAuth2 support  through the `quarkus-oidc` extension, which supports both xref:security-oidc-code-flow-authentication.adoc[Authorization Code Flow] and xref:security-oidc-bearer-token-authentication.adoc[Bearer token] authentication mechanisms.
```

sberyozkin replied:
I'm not sure this one works. I'd like to start with 1 sentence overview of what Quarkus OIDC can do

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,587 @@
+
+Well-known OIDC providers such as https://www.keycloak.org/documentation[Keycloak], https://developer.okta.com/[Okta], https://auth0.com/docs/[Auth0] and many xref:security-openid-connect-providers.adoc[social providers] can be configured with ease.
+
+In this tutorial you will learn how https://auth0.com/docs/[Auth0] can be used to secure your Quarkus endpoints.
```
Avoid using self-referential statements where possible:
"In this... "

Also, I think line 11 would make a great opening statement, especially on the landing page tile for skimmers:

```suggestion
Learn how to use the Quarkus Security OpenID Connect extension together with https://auth0.com/docs/[Auth0] authorization to secure your API endpoints.
```

sberyozkin replied:
This line is a more specific statement, we start with saying Quarkus can support all sort of providers and then say that here we'd look at one of them in particular.

But I can try to use your better worded sentence indeed, with a tiny update:

```Learn how to use the Quarkus OpenID Connect extension (`quarkus-oidc`) together with https://auth0.com/docs/[Auth0] OIDC provider to protect your API endpoints.`

Let me try and see if you will like it

sberyozkin commented:
Hi @michelle-purcell Thanks for starting a review, I'm just about to push an update related to the integration testing, but I'll pick up the suggestions you have already made too

sberyozkin commented:
@michelle-purcell I've done a couple of minor updates based on your feedback so far, please have a look tomorrow or so, thanks

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,743 @@
+
+xref:security-architecture.adoc[Quarkus Security] provides comprehensive OpenId Connect (OIDC) and OAuth2 support with its `quarkus-oidc` extension, supporting both xref:security-oidc-code-flow-authentication.adoc[Authorization code flow] and xref:security-oidc-bearer-token-authentication.adoc[Bearer token] authentication mechanisms.
+
+Well-known OIDC providers such as https://www.keycloak.org/documentation[Keycloak], https://developer.okta.com/[Okta], https://auth0.com/docs/[Auth0] and many xref:security-openid-connect-providers.adoc[social providers] can be configured with ease.
```
```suggestion
With Quarkus, you can easily configure xref:security-openid-connect-providers.adoc[well-known social OIDC and OAuth2 providers], including link:https://www.keycloak.org/documentation[Keycloak], link:https://developer.okta.com/[Okta], and link:https://auth0.com/docs/[Auth0].
```

sberyozkin replied:
Hi Michelle, would you be OK if I tried:
```
With Quarkus, you can easily configure OIDC providers such as link:https://www.keycloak.org/documentation[Keycloak], link:https://developer.okta.com/[Okta], and link:https://auth0.com/docs/[Auth0], and xref:security-openid-connect-providers.adoc[well-known social OIDC and OAuth2 providers]
```
?

See, `Keycloak`, `Okta` and `Auth0`, these are not qualified as `social` providers, they are `enterprise` grade (though I would not use `enterprise` here), one thing is certain is that when you see social login options, they'd never offer Keycloak or Okta or Autho :-), while xref:security-openid-connect-providers.adoc (somewhat misnamed file name) lists `social` providers only.

michelle-purcell replied:
@sberyozkin -  Your suggestion is perfect +1. I see what you mean. Thanks for explaining. Apologies for the churn.

sberyozkin replied:
@michelle-purcell Np at all, thanks for trying to improve the text

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,743 @@
+
+== Create Auth0 application
+
+Start with creating an `Auth0` application in the `Auth0` dashboard, for example, lets create a `QuarkusAuth0` web application:
```
```suggestion
Start with creating an `Auth0` application in the `Auth0` dashboard, for example, let's create a `QuarkusAuth0` web application:
```
But if following the Quarkus template numbered steps approach:

```suggestion
. Go to the `Auth0` dashboard and create a regular web application. For example, create an `Auth0` application called `QuarkusAuth0`.
```

@sberyozkin - Should I pause suggesting edits to convert the actual steps into numbered sub-steps in this review iteration?   Thanks.

sberyozkin replied:
@michelle-purcell I'm applying numbering scheme suggestions, but I'm not sure I'm following where to put `. ` next :-), please propose a few suggestions :-)

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,743 @@
+
+Start with creating an `Auth0` application in the `Auth0` dashboard, for example, lets create a `QuarkusAuth0` web application:
+
+image::auth0-create-application.png[Create Auth0 application]
```
@sberyozkin - Just some general comments about the screencaps instructional content.

- Screencaps are high maintenance. We typically try to avoid them for both accessibility and localization reasons.
- If the UIs in those screencaps get updated regularly and the docs don't get updated, we risk losing the trust of your reader or creating mismatch confusion.
- I would suggest you use them as judiciously as possible and ensure alt text is unique and descriptive to support visually impaired users.

michelle-purcell replied:
@sberyozkin - In saying that, _personally as a user/reader, I love them!_

sberyozkin replied:
Hi @michelle-purcell I'm not exactly sure I understand, when you say `screencaps`, do you mean the images ? The chance of the text in them changing is very low, if it does it won't change a lot, I've been hoping to  make sure the new Auth0 users who may not be well familiar with the Auth0 setup steps, just follow this tutorial.
We also hope this tutorial be reviewed externally, once published. Ideally, it will be very easy for the reviewers to follow.

Does it clarify why I've added (more than usual) number of images ?

sberyozkin replied:
I guess, since you liked it yourself, we can resolve this one :-) ?

sberyozkin replied:
@michelle-purcell Let me resolve just to optimize the review space but please re-open if I've missed something

michelle-purcell replied:
@sberyozkin - Thanks for explaining the rationale here. 

Yes, I mean the screen capture images of the Auth0 UI and other UIs, especially the UI that we don't own. 

If you think they are definitely required and the maintenance risks are low then keep them. 

Important: The screen captures of the UI must be supplemental only for the novice user. So the key question is:
If you removed any of the UI images, would the user have enough information with the text alone to follow and complete the required actions? 

If the answer is no, then we need another iteration to enhance the textual instructions for accessibility support.

For every image in the guide, we also need to ensure that every image has an Alt text attribute that clearly describes the picture for someone who cannot see it. I can do the ALT text review of images after this initial iteration when the step text edits/comments/suggestions are closed out.
Make sense? Thanks.

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,743 @@
+
+image::auth0-create-application.png[Create Auth0 application]
+
+`Auth0` will create it and generate the client id and secret and show its HTTPS-based domain:
```
```suggestion
.Result
Your`Auth0` application gets created with a client ID, secret, and HTTPS-based domain.
```

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,743 @@
+
+image::auth0-created-application.png[Created Auth0 application]
+
+You will use these 3 properties to configure Quarkus at the next step.
```
```suggestion
```

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,743 @@
+
+image::auth0-add-user.png[Add Auth0 application users]
+
+Now that the initial `Auth0` application setup has been completed, you are ready to start creating and configuring a Quarkus endpoint. We will be updating the `Auth0` application configuration along the way as necessary.
```
```suggestion
Now that you have successfully created and configured your `Auth0` application, you are ready to start creating and configuring a Quarkus endpoint. 
In the steps that follow, you will continue to configure and update your `Auth0` application as needed.
```

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,743 @@
+
+Now that the initial `Auth0` application setup has been completed, you are ready to start creating and configuring a Quarkus endpoint. We will be updating the `Auth0` application configuration along the way as necessary.
+
+== Create Quarkus application
```
```suggestion
== Create a Quarkus application
```

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,743 @@
+
+== Create Quarkus application
+
+Use the following Maven command to create a Quarkus REST application which can be secured with its OIDC extension.
```
Specificity: Should REST be RESTEasy Reactive?

```suggestion
Use the following Maven command to create a Quarkus RESTEasy Reactive application that can be secured with the  Quarkus OIDC extension.
```

sberyozkin replied:
Sounds good

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,743 @@
+:create-app-extensions: resteasy-reactive,oidc
+include::{includes}/devtools/create-app.adoc[]
+
+Create the application workspace and import it into your favourite IDE, note a `GreetingResource` JAX-RS resource that will need to be secured:
```
US spelling.
- Consider rephrasing this sentence. I'm not quite sure what the note is about here.
- Consider using the admonition markup for notes.

```suggestion
Create the application workspace and import it into your favorite IDE.
Note: <rephrase the`GreetingResource` JAX-RS resource sentence>
```

sberyozkin replied:
@michelle-purcell I did not mean it to be a `Note`, just wanted to let users know that `GreetingResource` was generated,  which I guess does not warrant a Note-level highlight, can we just replace `note` with some other verb/expression ?

michelle-purcell replied:
@sberyozkin - I see :+1:  In that case, something like...

```suggestion
Create the application workspace and import it into your favorite IDE.
A `GreetingResource` JAX-RS resource gets generated, which you will need to secure.
```

Not sure if my suggestion here ^^ is technically correct. These questions might help to enhance it further:

- Which step in this procedure section actually generates a `GreetingResource` ?
- When do you need to secure it? 
- Will the reader know how to secure it? (or is it covered in the code sample that follows?)

Thanks

sberyozkin replied:
@michelle-purcell Sounds good, yeah, then we'll work in the tutorial to get it secured

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,743 @@
+}
+----
+
+Now request that only authenticated users can access the `hello()` method and update it to return the current user's name instead:
```
If this doesn't break the technical meaning...

```suggestion
Update the Quarkus application code so that only authenticated users can access the `hello()` method and update it to return the current username instead, as shown in the following example:
```

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,743 @@
+    }
+}
+----
+<1> Injected `JsonWebToken` (JWT) token has an `@IdToken` qualifier. `IdToken` represents a user authentication in the OIDC Authorization code flow.
```
```suggestion
<1> The injected `JsonWebToken` (JWT) token has an `@IdToken` qualifier.
`IdToken` represents the user authentication part of the OIDC authorization code flow mechanism.
```

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,743 @@
+<1> Injected `JsonWebToken` (JWT) token has an `@IdToken` qualifier. `IdToken` represents a user authentication in the OIDC Authorization code flow.
+<2> `io.quarkus.security.Authenticated` annotation is added to the `hello()` method requiring that only authenticated users can access it.
+
+Note that the access token acquited during the authorization code flow, alongside the ID token, is not used directly by the endpoint but is only used to access downstream services on behalf of the currently authenticated user. We will talk about using the access tokens later in this tutorial.
```
```suggestion
[NOTE]
====
The access token acquired during the authorization code flow, alongside the ID token, is not used directly by the endpoint but is only used to access downstream services on behalf of the currently authenticated user. 
More to come on the topic of "access tokens", later in this tutorial.
```

sberyozkin commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,743 @@
+
+You will use these 3 properties to configure Quarkus at the next step.
+
+Next, add users to this application by selecting the `Users` option in the application's dashboard, for example:
```
@michelle-purcell The step is indeed critical and good point, I suppose I'm hand-holding the reader :-), but I really would like to avoid any problems at all for not too experienced Auth0 users. But indeed, your suggestion is simpler and it is really obvious what to do after reading it :-). Let me apply

sberyozkin commented:
Hi @michelle-purcell Thanks for spending the time today on the review, I've applied most suggestions, resolved a couple of them without changing anything (please see comments, reopen if you'd like).
Added the numbering scheme instructions as proposed but I suspect more updates will be needed now :-)

Added the Prerequisites section but stopped short of adding `What you will build` - I typed it and realized it was nearly a copy and paste of the `Learn ...` sentence from the preamble. I guess we can avoid it.

Lets continue this week, if you'll have some time

Cheers

michelle-purcell commented:
> Hi @michelle-purcell Thanks for spending the time today on the review, I've applied most suggestions, resolved a couple of them without changing anything (please see comments, reopen if you'd like). Added the numbering scheme instructions as proposed but I suspect more updates will be needed now :-)
> 
> Added the Prerequisites section but stopped short of adding `What you will build` - I typed it and realized it was nearly a copy and paste of the `Learn ...` sentence from the preamble. I guess we can avoid it.
> 
> Lets continue this week, if you'll have some time
> 
> Cheers

Thanks @sberyozkin :+1: -  The `What you will build` section aims to give the reader much more detail about the final goal and outcome, including any specific artifacts that get generated to support the use case(s) described. We can omit/defer this if you think it's not needed. It also overlaps a bit with the summary section, which recaps but directs you to the next stop in the journey.

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,761 @@
+
+== Create a Quarkus application
+
+Use the following Maven command to create a Quarkus RESTEasy Reactive application that can be secured with the  Quarkus OIDC extension.
```
Remove extra space, which I introduced :see_no_evil: 

```suggestion
Use the following Maven command to create a Quarkus RESTEasy Reactive application that can be secured with the Quarkus OIDC extension.
```

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,761 @@
+quarkus.oidc.credentials.secret=${client-secret}
+----
+
+Here, you have configured Quarkus to use the created `Auth0` application's domain, client id and secret. `quarkus.oidc.application-type=web-app` is an indication to Quarkus that the OIDC Authorization code flow must be used (we'll consider the alternatives later in this tutorial).
```
```suggestion
In completing this step, you have just configured Quarkus to use the domain, client ID, and secret of the `Auth0` application that you created earlier. 
Setting the property `quarkus.oidc.application-type=web-app` instructs Quarkus to use the OIDC authorization code flow but there are also other methods, which are discussed later on in the tutorial.
```

sberyozkin replied:
@michelle-purcell `In completing this step` definitely reads better than `Here` :-)

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,761 @@
+
+Here, you have configured Quarkus to use the created `Auth0` application's domain, client id and secret. `quarkus.oidc.application-type=web-app` is an indication to Quarkus that the OIDC Authorization code flow must be used (we'll consider the alternatives later in this tutorial).
+
+Note that you will use Quarkus `devmode` to test the endpoint so the endpoint address will be `http://localhost:8080/hello`. You need to register this address as an allowed callback URL in your `Auth0` application:
```
```suggestion
To test the endpoint by using Quarkus dev mode, you must complete the following steps:
* Set the endpoint address to \http://localhost:8080/hello. 
* Register \http://localhost:8080/hello as an allowed callback URL in your `Auth0` application.
```

In AsciiDoc, you can use a backslash (\) char to escape URLs so that they don't become active links. The \ isn't rendered in the output.

sberyozkin replied:
```suggestion
The endpoint address will be  \http://localhost:8080/hello and you must register this address as an allowed callback URL in your `Auth0` application.
```

@michelle-purcell I've tried to simplify - I put quite a bit irrelevant content earlier overloading a simple instruction (ex. the same address will be in the prod mode)

michelle-purcell replied:
@sberyozkin - I see. How about this minor tweak to your suggested edits?

```suggestion
By default, the endpoint address for the application is set to  \http://localhost:8080/hello so you must register this address as an allowed callback URL in your `Auth0` application.
```

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,761 @@
+
+image::auth0-allowed-callback.png[Auth0 allowed callback URL]
+
+This means that you will access the Quarkus `http://localhost:8080/hello` endpoint from the browser and `Auth0` will redirect you back to the same address after the authentication has been completed.
```
```suggestion
Now when you access the Quarkus \http://localhost:8080/hello endpoint from the browser, `Auth0` redirects you back to the same address after the authentication is completed.
```

michelle-purcell commented:
@sberyozkin - Re ... https://github.com/quarkusio/quarkus/pull/35931#issuecomment-1736349280 

Sounds good :+1:  I made some further suggestions and will keep going to the middle-end section later today (after critical meetings are done).   Thanks.

sberyozkin commented:
Thanks @michelle-purcell for more comments, absolutely, please take your time, thanks

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,771 @@
+[id="security-oidc-auth0-tutorial"]
+= Protect Quarkus web application by using `Auth0` OpenID Connect provider
```
Missing article. 

```suggestion
= Protect Quarkus web application by using an `Auth0` OpenID Connect provider
```

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,771 @@
+
+== Prerequisites
+
+Reviewing https://auth0.com/docs/[Auth0 documentation] is recommended.
```
```suggestion
Before you begin, review the link:https://auth0.com/docs/[Auth0] documentation.
```

michelle-purcell replied:
Merge this into 16 as per suggestion below.
```suggestion
```

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,771 @@
+
+Reviewing https://auth0.com/docs/[Auth0 documentation] is recommended.
+
+Reviewing Quarkus OIDC documents about xref:security-oidc-code-flow-authentication.adoc[Authorization code flow] and xref:security-oidc-bearer-token-authentication.adoc[Bearer token] authentication mechanisms will make it easier for you to complete this tutorial.
```
How about merging >15 and >16 into one list of recommended reading?

```suggestion
To make it easier for you to complete this tutorial, review the following documentation before you begin:

* link:https://auth0.com/docs/[Auth0 docs site]
* xref:security-oidc-code-flow-authentication.adoc[Quarkus OpenID Connect authorization code flow mechanism for protecting web applications]
* xref:security-oidc-bearer-token-authentication.adoc[Quarkus OpenID Connect (OIDC) Bearer token authentication]
```

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,771 @@
+
+== Create an Auth0 application
+
+. Go to the `Auth0` dashboard and create a regular web application. For example, create an `Auth0` application called `QuarkusAuth0`.
```
Looks odd having only 1 numbered step in this procedure.

Either make all actions part of an ordered list with numbered steps in sequence or remove this lone numbered step.

The tutorial template encourages the numbering of all action steps in a procedure, however this approach isn't widely adopted by contributors from what I can see.

sberyozkin replied:
@michelle-purcell Right, I'm not I was understanding it would add numbers like `1.`, I'd just remove it

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,771 @@
+
+== Create an Auth0 application
+
+. Go to the `Auth0` dashboard and create a regular web application. For example, create an `Auth0` application called `QuarkusAuth0`.
```
```suggestion
. Go to the `Auth0` dashboard and create a regular web application. 
For example, create an `Auth0` application called `QuarkusAuth0`.
```

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,771 @@
+image::auth0-create-application.png[Create Auth0 application]
+
+.Result
+Your`Auth0` application gets created with a client ID, secret, and HTTPS-based domain.
```
Does Auth0 need to be inside backticks?
It's a valid proper noun. If Vale rejects it, I can fix that rule to remove the false positive.

```suggestion
Your Auth0 application gets created with a client ID, secret, and HTTPS-based domain.
```

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,771 @@
+
+image::auth0-created-application.png[Created Auth0 application]
+
+Next, while still in the `Auth0` dashboard, add some users to your application.
```
```suggestion
Next, while still in the Auth0 dashboard, add some users to your application.
```

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,771 @@
+
+[NOTE]
+====
+The access token acquired during the authorization code flow, alongside the ID token, is not used directly by the endpoint but is only used to access downstream services on behalf of the currently authenticated user. 
```
Misplaced only:

```suggestion
The access token acquired during the authorization code flow, alongside the ID token, is not used directly by the endpoint but is used only to access downstream services on behalf of the currently authenticated user. 
```

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,771 @@
+More to come on the topic of "access tokens", later in this tutorial.
+====
+
+Configure OIDC in `application.properties` with the properties from the created `Auth0` application:
```
As there are two platforms being used in this tutorial, would it help to set the context in places, for example:

```suggestion
Configure OIDC in the Quarkus `application.properties` file  by using the properties from the Auth0 application that you created earlier.
```

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,771 @@
+quarkus.oidc.credentials.secret=${client-secret}
+----
+
+In completing this step, you have just configured Quarkus to use the domain, client ID, and secret of the `Auth0` application that you created earlier. 
```
Same comment about backticks around Auth0:

```suggestion
In completing this step, you have just configured Quarkus to use the domain, client ID, and secret of yourAuth0 application.
```

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,771 @@
+----
+
+In completing this step, you have just configured Quarkus to use the domain, client ID, and secret of the `Auth0` application that you created earlier. 
+Setting the property `quarkus.oidc.application-type=web-app` instructs Quarkus to use the OIDC authorization code flow but there are also other methods, which are discussed later on in the tutorial.
```
```suggestion
Setting the property `quarkus.oidc.application-type=web-app` instructs Quarkus to use the OIDC authorization code flow, but there are also other methods, which are discussed later on in the tutorial.
```

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,771 @@
+In completing this step, you have just configured Quarkus to use the domain, client ID, and secret of the `Auth0` application that you created earlier. 
+Setting the property `quarkus.oidc.application-type=web-app` instructs Quarkus to use the OIDC authorization code flow but there are also other methods, which are discussed later on in the tutorial.
+
+The endpoint address will be  \http://localhost:8080/hello and you must register this address as an allowed callback URL in your `Auth0` application.
```
```suggestion
The endpoint address will be  \http://localhost:8080/hello, which must also be registered as a permitted callback URL in your Auth0 application.
```

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,771 @@
+
+image::auth0-allowed-callback.png[Auth0 allowed callback URL]
+
+Now when you access the Quarkus \http://localhost:8080/hello endpoint from the browser, `Auth0` redirects you back to the same address after the authentication is completed.
```
```suggestion
After completing this step. when you access the Quarkus \http://localhost:8080/hello endpoint from a browser, Auth0 redirects you back to the same address after the authentication is completed.
```

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,771 @@
+
+[NOTE]
+====
+Quarkus also allows to configure the callback path with the `quarkus.oidc.authentication.redirect-path` property. Why do you not have to configure a callback path for this Quarkus application ? By default, Quarkus will use the current request path as the callback path which is why you don't have to configure `quarkus.oidc.authentication.redirect-path=/hello`.
```
Could we simplify these 2 sentences:

```suggestion
By default, Quarkus automatically uses the current request path as the callback path.
But you can override the default behavior and configure a specific callback path by setting the Quarkus `quarkus.oidc.authentication.redirect-path` property, as outlined in the following example:

`quarkus.oidc.authentication.redirect-path=/hello`
```

sberyozkin commented:
Thanks @michelle-purcell for another round of suggestions, it is nearly there :-)

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,778 @@
+
+Now you are ready to start testing the endpoint.
+
+== Test Quarkus endpoint - first try
```
```suggestion
== Test the Quarkus endpoint
```

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,778 @@
+
+== Test Quarkus endpoint - first try
+
+Start Quarkus in `devmode`:
```
```suggestion
Start Quarkus in dev mode:
```

sberyozkin commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,778 @@
+
+`quarkus.oidc.authentication.redirect-path=/hello`
+
+In production, your application will most likely have a larger URL space, with many more endpoint addresses available. In such cases you will configure Quarkus with a dedicated callback (redirect) path, for example, `quarkus.oidc.authentication.redirect-path=/authenticated-welcome` and register this URL in the provider's dashboard as `Auth0` will not allow an open ended list of redirect URLs. Qua ...
```
Thanks @michelle-purcell. reads OK

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,778 @@
+$ mvn quarkus:dev
+----
+
+This is the only time you are expected to start Quarkus manually in devmode while working with this tutorial. All the configuration and code updates in the follow up sections is expected to be recognized by Quarkus without having to restart the application manually.
```
```suggestion
[NOTE]
====
This is the only time during this tutorial that you will need to manually start Quarkus in dev mode.
The configuration and code update steps in the remaining sections of this tutorial are automatically observed and processed by Quarkus without you needing to restart the application manually.
====
```

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,778 @@
+
+This is the only time you are expected to start Quarkus manually in devmode while working with this tutorial. All the configuration and code updates in the follow up sections is expected to be recognized by Quarkus without having to restart the application manually.
+
+Open the browser and access `http://localhost:8080/hello`, you will be redirected to `Auth0` where you will login:
```
```suggestion
Open the browser and access the following URL:
`http://localhost:8080/hello`
You will be redirected to Auth0 and prompted to  log in:
```

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,778 @@
+
+image::auth0-login.png[Auth0 Login]
+
+and authorize the `QuarkusAuth0` application to access your account:
```
```suggestion
Next, give the `QuarkusAuth0` application authorization to access your account:
```

sberyozkin replied:
@michelle-purcell It reads a little bit unusual in scope of describing the OIDC code flow, though it is surely a perfect text as is :-), I'll keep the original text if you don't mind - this is a typical combination AFAIK, `user authorizes` or `user grants permissions`

michelle-purcell replied:
:+ - One thing: fix capitalization of the sentence. How about:

```suggestion
Authorize the `QuarkusAuth0` application to access your account:
```

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,778 @@
+
+image::auth0-authorize.png[Auth0 Authorize]
+
+and finally you will be redirected back to the Quarkus endpoint and get a response, `Hello, auth0|60e5a305e8da5a006aef5471`.
```
```suggestion
Finally, you will be redirected back to the Quarkus endpoint with the following response:
`Hello, auth0|60e5a305e8da5a006aef5471`
```

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,778 @@
+
+This is not the current user's name though, so why has this strange sequence been returned ?
+
+Lets use https://quarkus.io/guides/security-openid-connect-dev-services#dev-ui-all-oidc-providers[Quarkus OIDC Dev UI console] to find out.
```
```suggestion
Let's use link:https://quarkus.io/guides/security-openid-connect-dev-services#dev-ui-all-oidc-providers[Quarkus OIDC Dev UI console] to find out.
```

michelle-purcell replied:
See suggestion above to replace this phrase:

```suggestion
```

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,778 @@
+
+and finally you will be redirected back to the Quarkus endpoint and get a response, `Hello, auth0|60e5a305e8da5a006aef5471`.
+
+This is not the current user's name though, so why has this strange sequence been returned ?
```
```suggestion
[TIP]
====
Notice that the current username does not get returned. 
To learn more about why this behavior occurs, see the xref:security-openid-connect-dev-services.adoc#dev-ui-all-oidc-providers[Dev UI for all OpenID Connect Providers] section of the "Dev Services and UI for OpenID Connect (OIDC)" guide.
====
```

michelle-purcell replied:
Replace with suggestion in line 185:

```suggestion
```

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,778 @@
+
+Lets use https://quarkus.io/guides/security-openid-connect-dev-services#dev-ui-all-oidc-providers[Quarkus OIDC Dev UI console] to find out.
+
+== Looking at Auth0 tokens in OIDC DevUI
```
```suggestion
== Looking at Auth0 tokens in the OIDC Dev UI
```

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,778 @@
+
+Quarkus provides a great https://quarkus.io/guides/dev-ui-v2[DevUI] experience. In particular, it offers an out of the box support for developing and testing OIDC endpoints with a Keycloak container, https://quarkus.io/guides/security-openid-connect-dev-services#dev-services-for-keycloak[DevService for Keycloak] which is launched when the OIDC provider's address is not already configured with `qu ...
+
+However, Quarkus users can continue using OIDC DevUI when the provider is already configured. Lets check it right now.
```
```suggestion
However, Quarkus users can continue using the OIDC Dev UI when the provider is already configured. 
Use the following instructions to check this configuration:
```

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,778 @@
+
+However, Quarkus users can continue using OIDC DevUI when the provider is already configured. Lets check it right now.
+
+First though, make sure you change your Quarkus application to have a `hybrid` application type, instead of the current `web-app`:
```
```suggestion
First, change your Quarkus application type from `web-app` to `hybrid`, as follows:
```

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,778 @@
+quarkus.oidc.client-id=sKQu1dXjHB6r0sra0Y1YCqBZKWXqCkly
+quarkus.oidc.credentials.secret=${client-secret}
+----
+<1> Application type is changed to `hybrid`. The reason it was changed to `hybrid` is because OIDC Dev UI currently supports an `SPA` mode only, i.e, it itself, using its own Java Script, authenticates users to the OIDC provider, and uses the access token as a `Bearer` token to access the Quarkus endpoint as a service.
```
```suggestion
<1> Application type is changed to `hybrid` because OIDC Dev UI currently supports `SPA` mode only.
For example, using its own Java Script, the application authenticates users to the OIDC provider and uses the access token as a Bearer token to access the Quarkus endpoint as a service.
```

sberyozkin replied:
@michelle-purcell It makes it not quite obvious which application uses `JavaScript`, let me tweak it a bit

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,778 @@
+----
+<1> Application type is changed to `hybrid`. The reason it was changed to `hybrid` is because OIDC Dev UI currently supports an `SPA` mode only, i.e, it itself, using its own Java Script, authenticates users to the OIDC provider, and uses the access token as a `Bearer` token to access the Quarkus endpoint as a service.
+
+Typically, Quarkus has to be configured with `quarkus.oidc.application-type=service` to support `Bearer` token authentication but it also supports a `hybrid` application type which means it can support both the Authorization code flow and Bearer token flows at the same time.
```
```suggestion
Typically, Quarkus must be configured with `quarkus.oidc.application-type=service` to support `Bearer` token authentication, but it also supports a `hybrid` application type, which means it can support both the Authorization code flow and Bearer token flows at the same time.
```

sberyozkin commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,778 @@
+
+== Looking at Auth0 tokens in OIDC DevUI
+
+Quarkus provides a great https://quarkus.io/guides/dev-ui-v2[DevUI] experience. In particular, it offers an out of the box support for developing and testing OIDC endpoints with a Keycloak container, https://quarkus.io/guides/security-openid-connect-dev-services#dev-services-for-keycloak[DevService for Keycloak] which is launched when the OIDC provider's address is not already configured with `qu ...
```
@michelle-purcell Sure, I was trying to use `xref` but forgot in this case (hopefully only in this case) :-)

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,778 @@
+
+Typically, Quarkus has to be configured with `quarkus.oidc.application-type=service` to support `Bearer` token authentication but it also supports a `hybrid` application type which means it can support both the Authorization code flow and Bearer token flows at the same time.
+
+The other thing you need to do is to configure the Auth0 application to allow the callbacks to OIDC Dev UI, its URL format is  `http://localhost:8080/q/dev-ui/io.quarkus.quarkus-oidc/${provider-name}-provider` where in this case the `${provider-name}` is `auth0`:
```
```suggestion
You also need to configure the Auth0 application to allow the callbacks to the OIDC Dev UI.
Use the following URL format:

`http://localhost:8080/q/dev-ui/io.quarkus.quarkus-oidc/${provider-name}-provider` 

* Where in this example, the `${provider-name}` is `auth0`
```

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,778 @@
+
+image::auth0-allowed-callbacks.png[Auth0 Allowed Callbacks]
+
+Now you are ready to use OIDC Dev UI with Auth0, open `http://localhost:8080/q/dev/` in the browser and you will find an OpenId Connect card linking to an `Auth0` provider SPA:
```
```suggestion

Now you are ready to use OIDC Dev UI with Auth0.

Open `http://localhost:8080/q/dev/` in a browser session. An OpenId Connect card that links to an Auth0 provider SPA displays, as follows:
```

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,778 @@
+
+image::auth0-devui.png[Auth0 DevUI]
+
+Click on the `Auth0 provider` link and press `Login into Single Page Application`:
```
```suggestion
Click *Auth0 provider* followed by *Login into Single Page Application*:
```

michelle-purcell commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,778 @@
+
+image::auth0-devui-login-to-spa.png[Auth0 DevUI Login to SPA]
+
+You will be redirected to `Auth0`, where you will login and be redirected to the OIDC Dev UI Dashboard:
```
```suggestion
You will be redirected to `Auth0`, where you will log in and be redirected to the OIDC Dev UI Dashboard:
```

michelle-purcell commented:
@sberyozkin - I completed the 2nd iteration review as far as line 210.

FYI:  I will be on PTO for the next 2 weeks. 
Please talk to @inoxx03 or the docs team if this is a priority that cannot wait for my return. Thanks :-)

sberyozkin commented:
Hi @michelle-purcell Thanks a million for your help and time, very appreciated, will ping Stefan, cheers !

sberyozkin commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,804 @@
+
+:create-app-artifact-id: quarkus-auth0
+:create-app-extensions: resteasy-reactive,oidc
+include::{includes}/devtools/create-app.adoc[]
```
This sequence creates a project without the generated code, so somehow it needs to be tweaked to have the starter code generated

sberyozkin commented on docs/src/main/asciidoc/security-oidc-auth0-tutorial.adoc:
```diff
@@ -0,0 +1,804 @@
+
+Now, clear the browser cookie cache just in case, open `http://localhost:8080/hello` again, authenticate to `Auth0` and get an expected user name.
+
+== Access Quarkus with Auth0 access tokens
```
Note somewhere that if the access token is also in the JWT format then it can be verified directly

sberyozkin commented:
@michalvavrik Hi Michal, it is entirely up to you and if you will have any time at all; if it can if of interest, then please confirm before next Wed (3.5.0.CR1) that the guide works precisely as described, and if something is confusing in the text and should be said somehow differently, please note that too. Np at all if you won't have time :-)

sberyozkin commented:
@michalvavrik Sorry, hope you did not even have time to consider checking this PR :-), I'll need to do more work around the access tokens/propagation section. In any case I already checked this tutorial twice, thanks

sberyozkin commented:
I've significantly reworked Access token and token propagation sections, showing Auth0 API, audience, scope to PermissionAllowed conversion. This draft sticks to the `echo:name` scope (I've used `echo.name` in tests due to #36313 - but it can be fixed easily early next week)

sberyozkin commented:
Hi @gastaldi - can you please consider giving this doc a read during the next few days ? Michelle did a very comprehensive review of may be up to 1/4 of the doc, but she is away and will likely have to be catching up with other urgent doc issues when she returns. I think we can iterate with the Doc team on keeping aligning the text with guidelines, and optimizing it. I wonder if fixing the obvious text typos or strangely expressed sentences would do for now, I'd not mind at all try to get to 3.5.0, the sooner it can become public for users to start working the it the better.
FYI, I've verified the actual steps twice, and can do a 3rd round once the review is complete.
Np at all if you won't find time, we'll get it over the line sooner or later :-)

sberyozkin commented:
I've enabled the review to get the HTML page -  I've just gone again through the whole text, tuned a few things, it looks quite ready to me.

gastaldi reviewed (approved):
LGTM

sberyozkin commented:
Hey @gastaldi, thanks for having a look :-). I'll do one more walk-through through the whole demo tomorrow or so (I only reread the text on Fri), as it really must work exactly as described, and will merge once I confirm it works. Cheers

sberyozkin commented:
Reacted to a few Lint suggestions

sberyozkin commented:
I've updated it more to make sure the integration tests pass when building in JVM and native modes, LGTM overall. I'll now merge it once the build gets green.

sberyozkin commented:
Hi @michelle-purcell, we can iterate and keep tuning the doc, thanks very much for the earlier review. I'd just like to try to get it to 3.5.0 and make it visible
